### DISCUSSION TITLE
Allowing sealed classes as generic constraints to help annotate nullability relationships

### DISCUSSION THREAD

**Discussion by safakgur** [dotnet/csharplang#3013] [Language Ideas]
In the following example, there is no way to annotate the nullability of _the returning Foo's_ generic argument to be the same as the nullability of _the accepted Foo's_ generic argument.

```c#
Foo<string?> Bar(Foo<string?> foo)
{
    if (foo.Value is null) { /* Something */ }
    else { /* Some other thing that uses the string API */ }
    return foo;
}

var f = new Foo<string>("s");
var r = Bar(f); // Warns because f's type parameter is not nullable
```

A signature like `Foo<T> Bar(Foo<T> foo)` would allow us to preserve the nullability and adding a type constraint like `where T : SomeClass?` would allow us to use type-specific functionality in the method. But sealed classes like string can't be used as type constraints.

I understand it didn't make sense before but I think it may be useful to lift this restriction (it is already allowed in IL) now with nullable reference types in the wild, so we can do:

```c#
Foo<T> Bar<T>(Foo<T> foo)
    where T : string?
{
    if (foo.Value is null) { /* Something */ }
    else { /* Some other thing that uses the string API */ }
    return foo;
}

var f1 = new Foo<string>("s");
var r1 = Bar(f1); // Foo<string>

var f2 = new Foo<string?>("s");
var r2 = Bar(f2); // Foo<string?>
```

I also [asked about this][1] on Stack Overflow.

[1]: https://stackoverflow.com/q/58300652/704144

**Comment by YairHalberstadt** [comment]
It's definitely an interesting problem, and the suggested solution does have a lot going for it.

However I'm not sure this would work in most cases. For example:

```csharp
Foo<T> Bar<T>(Foo<T> foo)
    where T : string?
{
    if (foo.Value is null) { return new Foo<T?>(null); }
    else { return new Foo<T>(foo.Value.Substring(1,2)); }
}
```

This will fail to compile, since the return type of `string.Substring` is `string`, not `T`.

**Comment by huoyaoyuan** [comment]
I think this will get help if `NotNullWhenNotNull` get extended.
